### PR TITLE
[SPARK-19549] Allow providing reason for stage/job cancelling

### PR DESCRIPTION
**Pull request by ala**
## What changes were proposed in this pull request?

This change add an optional argument to `SparkContext.cancelStage()` and `SparkContext.cancelJob()` functions, which allows the caller to provide exact reason  for the cancellation.

## How was this patch tested?

Adds unit test.

### CODE REVIEW

**Comment by ala** [comment]
@rxin

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -2207,20 +2207,22 @@ class SparkContext(config: SparkConf) extends Logging {
    * Cancel a given job if it's scheduled or running.
    *
    * @param jobId the job ID to cancel
+   * @param reason optional reason for cancellation
    * @note Throws `InterruptedException` if the cancel message cannot be sent
    */
-  def cancelJob(jobId: Int) {
-    dagScheduler.cancelJob(jobId)
+  def cancelJob(jobId: Int, reason: String = "") {
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: actually you need to define a new function with two parameters and keep the old one to maintain binary compatibility.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 2]
```diff
@@ -696,9 +696,9 @@ class DAGScheduler(
   /**
    * Cancel a job that is running or waiting in the queue.
    */
-  def cancelJob(jobId: Int): Unit = {
+  def cancelJob(jobId: Int, reason: String = ""): Unit = {
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: internally can we use an Option rather than empty string?

**Comment by rxin** [comment]
LGTM pending Jenkins.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 3]
```diff
@@ -2207,10 +2207,32 @@ class SparkContext(config: SparkConf) extends Logging {
    * Cancel a given job if it's scheduled or running.
    *
    * @param jobId the job ID to cancel
+   * @param reason optional reason for cancellation
    * @note Throws `InterruptedException` if the cancel message cannot be sent
    */
-  def cancelJob(jobId: Int) {
-    dagScheduler.cancelJob(jobId)
+  def cancelJob(jobId: Int, reason: String): Unit = {
+    dagScheduler.cancelJob(jobId, Some(reason))
```
Comment (hvanhovell) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Please use `Option(reason)`, `Some(reason)` will happily wrap a null.

**Comment by rxin** [comment]
Merging in master!

**Comment by hvanhovell** [comment]
retest this please

**Pull request closed without merging**